partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelExclude.kt

```diff
@@ -0,0 +1,178 @@
+            !attrsToRemove.contains(structField.first)
+        }
+        val finalStruct = entriesWithRemoved.map { structField ->
+            val name = structField.first
+            var expr = structField.second
+            // apply case-sensitive tuple attr exclusions
+            val structFieldCaseSensitiveKey = ExcludeStep.StructField(name, ExcludeFieldCase.SENSITIVE)
+            branches.find {
+                it.step == structFieldCaseSensitiveKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            // apply case-insensitive tuple attr exclusions
+            val structFieldCaseInsensitiveKey = ExcludeStep.StructField(name, ExcludeFieldCase.INSENSITIVE)
+            branches.find {
+                it.step == structFieldCaseInsensitiveKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            // apply tuple wildcard exclusions
+            val tupleWildcardKey = ExcludeStep.StructWildcard
+            branches.find {
+                it.step == tupleWildcardKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            Pair(name, expr)
+        }
+        return structValue(finalStruct)
+    }
+
+    /**
+     * Returns a [PartiQLValue] created from an iterable of [coll]. Requires [type] to be a collection type
+     * (i.e. [PartiQLValueType.LIST], [PartiQLValueType.BAG], or [PartiQLValueType.SEXP]).
+     */
+    @OptIn(PartiQLValueExperimental::class)
+    private fun newCollValue(type: PartiQLValueType, coll: Iterable<PartiQLValue>): PartiQLValue {
+        return when (type) {
+            PartiQLValueType.LIST -> listValue(coll)
+            PartiQLValueType.BAG -> bagValue(coll)
+            PartiQLValueType.SEXP -> sexpValue(coll)
+            else -> error("Collection type required")
+        }
+    }
+
+    @OptIn(PartiQLValueExperimental::class)
+    private fun excludeOnCollValue(
+        coll: CollectionValue<*>,
+        type: PartiQLValueType,
+        exclusions: ExcludeNode
+    ): PartiQLValue {
+        val leavesSteps = exclusions.leaves.map { leaf -> leaf.step }
+        val branches = exclusions.branches
+        if (leavesSteps.any { it is ExcludeStep.CollWildcard }) {
+            // collection wildcard at current level. return empty collection
+            return newCollValue(type, emptyList())
+        } else {
+            val indexesToRemove = leavesSteps.filterIsInstance<ExcludeStep.CollIndex>()
+                .map { it.index }
+                .toSet()
+            val collWithRemoved = when (coll) {
+                is BagValue -> coll
+                is ListValue, is SexpValue -> coll.filterIndexed { index, _ ->
+                    !indexesToRemove.contains(index)
+                }
+            }
+            val finalColl = collWithRemoved.mapIndexed { index, element ->
```

Comment:
what's the difference between collWithRemoved and finalColl? I think you should be applying the exclusions with a mapIndexedNotNull and return null if the index is skipped otherwise apply the exlcusion. This way you don't traverse the collection more than once.

Reply:
`collWithRemoved` would remove the indexes at that level (i.e. the leaf nodes with `ExcludeStep.CollIndex`). `finalColl` would remove the values at deeper nested levels (i.e. the branch nodes).

> I think you should be applying the exclusions with a mapIndexedNotNull and return null if the index is skipped otherwise apply the exlcusion. This way you don't traverse the collection more than once.

Good point. I will change to use `mapIndexedNotNull` so only one iteration of `coll` is necessary.